crates/red_knot_python_semantic/src/symbol.rs

```diff
@@ -67,6 +67,30 @@ impl<'db> Symbol<'db> {
             .expect("Expected a (possibly unbound) type, not an unbound symbol")
     }
 
+    /// Transform the symbol into a [`LookupResult`],
+    /// a [`Result`] type in which the `Ok` variant represents a definitely bound symbol
+    /// and the `Err` variant represents a symbol that is either definitely or possibly unbound.
+    pub(crate) fn into_lookup_result(self) -> LookupResult<'db> {
+        match self {
+            Symbol::Type(ty, Boundness::Bound) => Ok(ty),
+            Symbol::Type(ty, Boundness::PossiblyUnbound) => Err(LookupError::PossiblyUnbound(ty)),
+            Symbol::Unbound => Err(LookupError::Unbound),
+        }
+    }
```

Comment:
In my work attempting to do a wholesale refactor of our `Outcome` enums, I looked for a while at whether it might be possible to use `Result`s for all of them. I'm no longer convinced that doing so would be a good idea, but it's still the case that a lot of the time we want to treat these `Outcome`s "like `Result`s".

Reply:
...and for more on that topic, see https://github.com/astral-sh/ruff/pull/16161 😃